### PR TITLE
feat/add_analysis_route

### PR DESCRIPTION
**Pull request by beurie-v**
HOW TO TEST:

- SIGNUP: POST /api/signup BODY:
{
	"firstname": "",
	"lastname": "",
	"email": "",
	"password": ""
}
- LOGIN: POST /api/login BODY:
{
  "email": "",
  "password": ""
}

- SIMULATE DATA: 
GET /api/simulation?timestamp=XXX
GET /api/simulation/ill?timestamp=XXX

- START FIRST ANALYSIS: POST /api/analysis (Should return code: 2 -> model created)

- RE-SIMULATE DATA

- START ANALYSIS: POST /api/analysis (Should return code: 0 (No anomaly) or code: 1 (Anomaly detected)

### CODE REVIEW

File: server/server.js [hunk 1]
```diff
@@ -63,6 +63,7 @@ app.use(function (req, res, next) {
 });
 
 app.use(function errorHandler (err, req, res, next) {
+    console.log("[ERROR]: " + err.message);
```
Comment (MickaToast) [inline server/server.js]: console.error

File: server/src/MachineLearning/ml.js [hunk 2]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
-                .then(function (dataset) {
-                    if (dataset === null || dataset.UUID != UUID) {
-                        err.message = "Could not find any dataset with id: " + datasetId + " that belongs to that user"
-                        err.status = 404;
-                        reject(err);
-                    }
+        if (dataset.UUID != UUID) {
+            err.message = "Dataset " + dataset.id + " does not belong to that user";
+            err.status = 500;
+            reject(err);
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: reject does not end condition, you might need a return

File: server/src/MachineLearning/ml.js [hunk 3]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
+        }
 
-                    if (!validateAttributes(JSON.parse(model.attribute), JSON.parse(dataset.attribute))) {
-                        err.message = "Attributes do not match between model: " + model.id + " and dataset: " + dataset.id;
+        if (!validateAttributes(JSON.parse(model.attribute), JSON.parse(dataset.attribute))) {
+            err.message = "Attributes do not match between model: " + model.id + " and dataset: " + dataset.id;
+            err.status = 500;
+            reject(err);
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: reject does not end condition, you might need a return

File: server/src/MachineLearning/ml.js [hunk 4]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
+        var modelData = JSON.parse(model.data);
+        var datasetData = createData(dataset.attribute, dataset.data, dataset.id);
+        var anomaly = false;
+
+        Object.keys(modelData).forEach(function (sensorName) {
+            if (!anomaly) {
+                if (!datasetData[sensorName]) {
+                    err.message = "Inconsistency in sensor names between dataset " + dataset.id + " and model " + model.id;
+                    err.status = 500;
+                    reject(err);
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: reject does not end condition, you might need a return

File: server/src/MachineLearning/ml.js [hunk 5]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
                     }
-                })
-                .catch(function (err) {
-                    console.log("[ERROR] In classify: " + err.message);
-                    reject(err)
-                });
+                }
+            }
+        });
+        if (anomaly) {
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: Should be better to resolve the anomaly since it's a boolean

File: server/src/MachineLearning/ml.js [hunk 6]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
+                    var modelValue = parseFloat(modelData[sensorName]);
+                    var datasetValue = parseFloat(datasetData[sensorName]);
+                    if (isNaN(modelValue)) {
+                        err.message = "Found NaN value in model " + model.id + " data";
+                        err.status = 500;
+                        reject(err);
+                    } else if (isNaN(datasetValue)) {
+                        err.message = "Found NaN value in dataset " + dataset.id;
                         err.status = 500;
                         reject(err);
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: reject does not end condition, you might need a return

File: server/src/MachineLearning/ml.js [hunk 7]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
+                    err.message = "Inconsistency in sensor names between dataset " + dataset.id + " and model " + model.id;
+                    err.status = 500;
+                    reject(err);
+                } else {
+                    var modelValue = parseFloat(modelData[sensorName]);
+                    var datasetValue = parseFloat(datasetData[sensorName]);
+                    if (isNaN(modelValue)) {
+                        err.message = "Found NaN value in model " + model.id + " data";
+                        err.status = 500;
+                        reject(err);
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: reject does not end condition, you might need a return

File: server/src/MachineLearning/ml.js [hunk 8]
```diff
@@ -64,98 +63,84 @@ function validateAttributes(modelAttr, datasetAttr) {
...
-                .then(function (dataset) {
-                    if (dataset === null || dataset.UUID != UUID) {
-                        err.message = "Could not find any dataset with id: " + datasetId + " that belongs to that user"
-                        err.status = 404;
-                        reject(err);
-                    }
+        if (dataset.UUID != UUID) {
+            err.message = "Dataset " + dataset.id + " does not belong to that user";
+            err.status = 500;
+            reject(err);
```
Comment (MickaToast) [inline server/src/MachineLearning/ml.js]: reject does not end condition, you might need a return

File: server/server.js (hunk 1 above)
Comment (beurie-v) [reply server/server.js]: Fixed

File: server/src/MachineLearning/ml.js (hunk 2 above)
Comment (beurie-v) [reply server/src/MachineLearning/ml.js]: Fixed

File: server/src/MachineLearning/ml.js (hunk 4 above)
Comment (beurie-v) [reply server/src/MachineLearning/ml.js]: Fixed

File: server/src/MachineLearning/ml.js (hunk 3 above)
Comment (beurie-v) [reply server/src/MachineLearning/ml.js]: Fixed

File: server/src/MachineLearning/ml.js (hunk 6 above)
Comment (beurie-v) [reply server/src/MachineLearning/ml.js]: Fixed

File: server/src/MachineLearning/ml.js (hunk 8 above)
Comment (beurie-v) [reply server/src/MachineLearning/ml.js]: Fixed

File: server/src/MachineLearning/ml.js (hunk 7 above)
Comment (beurie-v) [reply server/src/MachineLearning/ml.js]: Fixed

**Pull request merged**